amm-executor/src/Spectrum/Executor/Backlog/Service.hs

```diff
@@ -1,60 +1,173 @@
 
-data Backlog m = Backlog
+data BacklogService m = BacklogService
   { put        :: OrderInState 'Pending -> m ()
   , suspend    :: OrderInState 'Suspended -> m Bool
   , checkLater :: OrderInState 'InProgress -> m Bool
   , tryAcquire :: m (Maybe Order)
-  , drop       :: OrderId -> m Bool
+  , drop       :: OrderId -> m ()
   }
 
-mkBacklog
-  :: forall f m. (MonadIO f, MonadResource f, MonadIO m, MonadThrow m)
+mkBacklogService
+  :: forall f m. (MonadIO f, MonadIO m, LiftK m f)
   => MakeLogging f m
-  -> BacklogConfig
-  -> f (Backlog m)
-mkBacklog MakeLogging{..} BacklogConfig{..} = do
-  logging     <- forComponent "Pools"
+  -> BacklogServiceConfig
+  -> BacklogStore m
+  -> f (BacklogService m)
+mkBacklogService MakeLogging{..} config store@BacklogStore{..} = do
+  logging     <- forComponent "BacklogService"
   -- those queues should be shared with Backlog.Proceess (to make live updates). So maybe worth extracting them into separate module. e.g. BacklogStore
   pendingPQ   <- newIORef PQ.empty  -- ordered by weight; new orders
   suspendedPQ <- newIORef PQ.empty  -- ordered by weight; failed orders, waiting for retry (retries are performed with some constant probability, e.g. 5%) 
   toRevisitQ  <- newIORef Seq.empty -- regular queue; successully submitted orders. Left orders should be re-executed in X minutes. Normally successfully confirmed orders are eliminated from this queue.
+  _           <- liftK $ recover store config pendingPQ
+  pure $ attachLogging logging BacklogService
+    { put = \(PendingOrder order timestamp) -> do
+        put $ BacklogOrder timestamp order
+        modifyIORef pendingPQ (PQ.insert $ mkWeightedOrderWithTimestamp order timestamp)
+    , suspend = \(SuspendedOrder order timestamp) -> do
+        modifyIORef suspendedPQ (PQ.insert $ mkWeightedOrderWithTimestamp order timestamp)
+        exists $ BacklogOrder timestamp order
+    , checkLater = \(InProgressOrder order timestamp) -> do
+        modifyIORef toRevisitQ (mkWeightedOrderWithTimestamp order timestamp Seq.<|)
+        exists $ BacklogOrder timestamp order
+    , tryAcquire = do
+        revisitOrders config store pendingPQ toRevisitQ
+        getMaxWeightedOrder' config store pendingPQ suspendedPQ
+    , drop = dropOrder
+    }
+
+revisitOrders
+  :: (MonadIO m)
+  => BacklogServiceConfig
+  -> BacklogStore m
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> IORef (Seq.Seq WeightedOrderWithTimestamp)
+  -> m ()
+revisitOrders BacklogServiceConfig{..} BacklogStore{..} pendingQueueRef toRevisitSeqRef = do
+  currentTime <- getCurrentTime
+  revisited2process <-
+    atomicModifyIORef'
+      toRevisitSeqRef
+      (Seq.spanl (\(WeightedOrderWithTimestamp _ _ oTime) -> diffUTCTime currentTime oTime < orderExecTime))
+  mapM_ (\order@(WeightedOrderWithTimestamp oId _ oTime) ->
+          if diffUTCTime currentTime oTime < orderLifetime
+            then modifyIORef' pendingQueueRef (PQ.insert order)
```

Comment:
Why here are you using strict version of modifyIORef?

Reply:
revert to non-strict

Reply:
I asked not to revert to non-strict. I'm asking why are you choosing one over another?